File: cpp/templates/Datastructure/FenwickTree.cpp
#include <bits/stdc++.h>
using namespace std;
template <typename T>
class FenwickTree {
public:
	int n;
	std::vector<T> tree;
	FenwickTree(int size) : n(size + 2), tree(size + 2, 0) {}
	void update(int ind, int delta) {
		for (++ind; ind < n; ind += ind & -ind)
			tree[ind] += delta;
	}
	T prefix_query(int ind) const {
		T ans = 0;
		for (++ind; ind > 0; ind -= ind & -ind)
			ans += tree[ind];
		return ans;
	}
};

template <typename T>
class RangeFenwick {
private:
	int n;
	std::vector<T> tree1, tree2;
	void update(std::vector<T> & tree, int idx, T delta) {
		idx += 1; 
		while (idx <= n) {
			tree[idx] += delta;
			idx += idx & -idx;
		}
	}
	T query(const std::vector<T> & tree, int idx) const {
		T res = 0;
		while (idx > 0) {
			res += tree[idx];
			idx -= idx & -idx;
		}
		return res;
	}
public:
	RangeFenwick(int size) : n(size), tree1(size + 1, 0), tree2(size + 1, 0) {}
	void range_update(int l, int r, T delta) {
		update(tree1, l, delta);
		update(tree1, r + 1, -delta);
		update(tree2, l, delta * l);
		update(tree2, r + 1, -delta * (r + 1));
	}
	int prefix_query(int idx) const {
		idx += 1;
		return query(tree1, idx) * idx - query(tree2, idx);
	}
};

template<typename T>
class Fenwick2D {
  T n, m;
  vector<vector<T>> tree;
public:
  Fenwick2D(T n, T m) : n(n), m(m) {
    tree.resize(n + 1, vector<T>(m + 1, 0));
  }
  void update(T x, T y, T delta) {
    while (x <= n) {
      T y1 = y;
      while (y1 <= m) {
        tree[x][y1] += delta;
        y1 += y1 & -y1;
      }
      x += x & -x;
    }
  }
  
  T query(T x, T y) {
    T result = 0;
    while (x > 0) {
      T y1 = y;
      while (y1 > 0) {
        result += tree[x][y1];
        y1 -= y1 & -y1;
      }
      x -= x & -x;
    }
    return result;
  }
  T rectangle_sum(T x1, T y1, T x2, T y2) {
    x1 += 1, y1 += 1, x2 += 1, y2 += 1;
    return query(x2, y2) - query(x1 - 1, y2) - query(x2, y1 - 1) + query(x1 - 1, y1 - 1);
  }
};

int main() {
	FenwickTree<int> ft(10);
	ft.update(1, 1);
	ft.update(2, 2);
	ft.update(3, 3);
	cout << ft.prefix_query(6) << endl;
	return 0;
}

File: python/DataStructure/FenwickTree.py
class FenwickTree:
  def __init__(self, n):
    self.n, self.tree = n + 2, [0] * (n + 2)
  
  def add(self, ind, x):
    ind += 1
    while ind <= self.n:
      self.tree[ind] += x
      ind += (ind & -ind)
  
  def prefix_sum(self, ind):
    ind, res = ind + 1, 0
    while ind > 0:
      res += self.tree[ind]
      ind -= (ind & -ind)
    return res

  
class RangeFenwick:
  def __init__(self, size):
    self.n = size
    self.tree1 = [0] * (self.n + 1) 
    self.tree2 = [0] * (self.n + 1) 

  def _update(self, tree, idx, delta):
    idx += 1
    while idx <= self.n:
      tree[idx] += delta
      idx += idx & -idx

  def _query(self, tree, idx):
    idx, res = idx + 1, 0
    while idx > 0:
      res += tree[idx]
      idx -= idx & -idx
    return res

  def update(self, l, r, delta):
    self._update(self.tree1, l, delta)
    self._update(self.tree1, r + 1, -delta)
    self._update(self.tree2, l, delta * l)
    self._update(self.tree2, r + 1, -delta * (r + 1))

  def prefix_query(self, idx):
    return self._query(self.tree1, idx) * (idx + 1) - self._query(self.tree2, idx)
  
class Fenwick2D:
  def __init__(self, n, m):
    self.n, self.m = n, m
    self.tree = [[0] * (m + 1) for _ in range(n + 1)]

  def update(self, x, y, delta):
    xi = x
    while xi <= self.n:
      yi = y
      while yi <= self.m:
        self.tree[xi][yi] += delta
        yi += yi & -yi
      xi += xi & -xi

  def query(self, x, y):
    result, xi = 0, x
    while xi > 0:
      yi = y
      while yi > 0:
        result += self.tree[xi][yi]
        yi -= yi & -yi
      xi -= xi & -xi
    return result

  def rectangle_sum(self, x1, y1, x2, y2):
    return self.query(x2, y2) - self.query(x1 - 1, y2) - self.query(x2, y1 - 1) + self.query(x1 - 1, y1 - 1)
